### ISSUE TITLE
Not merging mongoose Models

### ISSUE THREAD

**Issue by sansSpoon** [KyleAMathews/deepmerge#99]
I'd rather be using this lightweight library to handle deep merges instead of lodash. However after several attempts I'm unable to do a merge (even for a top level key) on a [mongoose](http://mongoosejs.com/docs/api.html#Model) model object.

```js
const merge = require('deepmerge');

var modelToUpdate = {"employees":[{"_id":"5a97e47097db2746fb986e8c","firstName":"John","lastName":"Doe","__v":0}],"_id":"5a98950568b7564cacc60c85","companyName":"Edit this top level","__v":0}

var updatePath = {companyName: "Not working"}

merge(modelToUpdate, updatePath)

// => {"employees":[{"_id":"5a97e47097db2746fb986e8c","firstName":"John","lastName":"Doe","__v":0}],"_id":"5a98950568b7564cacc60c85","companyName":"Edit this top level","__v":0}
```

If I try this with either `Object.assign()` or `lodash.merge()` I get the correct output:

```js
 {"employees":[{"_id":"5a97e47097db2746fb986e8c","firstName":"John","lastName":"Doe","__v":0}],"_id":"5a98950568b7564cacc60c85","companyName":"Not working","__v":0}
```
I've json.stringified the results of the object being merged for clarity in this post.
Has anybody got any experience with handling Mongoose models?

**Comment by macdja38** [comment]
Just to confirm, are you checking to see if modelToUpdate is being modified? or are you using the result of merge()?

**Comment by sansSpoon** [comment]
Thanks for the quick reply! Here's my full route methods:

```js
exports.params = async function companyParams(id, ctx, next) {
		ctx.company = await Company.findById(id).populate('employees').exec();
		if (!ctx.company) {
			return ctx.status = 404;
		}
		return next();
}

exports.updateOne = async function updateOne (ctx, next) {
	merge(ctx.company, ctx.request.body);
	ctx.company.save(function(err, result){
		if (err) {
			next(err);
		} else {
			ctx.body = result;
		}
	});
	await next();
}
```
As I mentioned, If I swap out `merge(ctx.company, ctx.request.body)` for `_.merge(ctx.company, ctx.request.body)`, I get good results.

**Comment by sansSpoon** [comment]
Thanks again for the further feedback. I've reworked the `updateOne` method so that the `.save` is applied to the newly created object. I'm now getting an error. I added in some logging to see what is happening with the new object. As you can see, It's not the same as the original, missing the `employees` array, `__v` property, and most importantly the `_id: ObjectID` property. It's also dropped the "model" name from the object listing in the console, not really sure what that's all about—excuse my ignorance.

```js
exports.updateOne = async function updateOne (ctx, next) {
	
	console.log("ctx.company");
	console.log(ctx.company);
	
	console.log("ctx.request.body");
	console.log(ctx.request.body);
	
	console.log("Merging with deepmerge");
	let myMerged = merge(ctx.company, ctx.request.body);
	console.log(myMerged);
	
	myMerged.save(function(err, result){
		if (err) {
			next(err);
		} else {
			console.log(result);
			ctx.body = result;
		}
	});
	await next();
}
```

![screen shot 2018-03-02 at 7 09 25 pm](https://user-images.githubusercontent.com/22439699/36889478-3c388fa0-1e4e-11e8-8523-5cba6aea5da1.png)

**Comment by TehShrike** [comment]
I don't see `save` in either of the functions that you're merging there.

I think `deepmerge` is working fine - in that first example, you were running into difficulty because you were expecting `deepmerge` to modify the first object that you pass in to it (like `Object.assign`), but it doesn't - it makes a new object and copies properties onto it.

**Comment by macdja38** [comment]
Is the save function on a prototype or an actual property of the ctx.company?

**Comment by sansSpoon** [comment]
@TehShrike, yep, the first instance I wan't using `deepmerge` correctly. I fixed that in the 2nd iteration and applied a `.save` to the object that `deepmerge` created:

```js
myMerged.save(function(err, result){
	if (err) {
		next(err);
	} else {
		console.log(result);
		ctx.body = result;
	}
});
```
Deepmerge is working by merging the `ctx.request.body` onto the `ctx.company`, in this case the company name has changed. However the new `myMerged` object that `deepmerge` has created no longer has all the properties of the original, specifically `_id: ObjectID`. The `myMerged.save()` is throwing the error because of this.

@macdja38, I think you've zeroed in on the issue there. from the docs. `.save()` is `Model.prototype.save()`. If this is a new object that isn't an instance of `Model`, then it's not going to save.

So the question becomes, is there any way to turn this new object `myMerged`, back into an instance of `Model`?

**Comment by sansSpoon** [comment]
I tried setting the `myMerged` object's constructor back to Model with:

```js
let Company = mongoose.model('company', companySchema);
myMerged.constructor = Company;
```

No luck. Though I think this is now outside the scope of `deepmerge`.

**Comment by TehShrike** [comment]
@sansSpoon check out https://github.com/KyleAMathews/deepmerge/pull/96 - I should probably merge and deploy it.  I suspect that if you used that version, following the `is-plain-object` example, it would behave the way you want.

**Comment by TehShrike** [comment]
@sansSpoon the custom `isMergeableObject` option is released under 2.1.0.

If you use the `is-plain-object` example, I believe you will get the behavior you expect: https://github.com/KyleAMathews/deepmerge#ismergeableobject

**Comment by devmotheg** [comment]
Sadly this is still an issue.
